### DISCUSSION TITLE
Infer type of a list with subclass objects?

### DISCUSSION THREAD

**Discussion by diepala** [microsoft/pyright#3616] [Ideas]
Currently, if a list (or other collections) have objects with different types, pyright infers that the type of the list is `list[Any]`. But wouldn't it be better if it could use the largest common base class?

Take this example:
```python
class BaseClass():
    pass

class A(BaseClass):
    pass

class B(BaseClass):
    pass

var = [A(), B()]
```

With the current implementation, the inferred type of `var` is `list[Any]`, but wouldn't it be better if it was `list[BaseClass]`?

I know pyright has the option to infer narrower types, but when we set strictListInference to True, then the inferred type of `var` is `list[A | B]`. It would be nice to have the intermediate option.

Also, the strictListInference is not well-supported with dictionaries. For example, the inferred type of `dic = {"a": A(), "b": B()}` is `dict[str, Any]`, even when strictListInference is set to True.

**Comment by erictraut** [comment]
Pyright's inference rules for lists are documented [here](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md#list-expressions). We've tuned these rules over time based on real-world use cases. There's no "right" answer for type inference, and every approach has pros and cons. We think that our current inference rules strike a good balance. We're unlikely to change them at this point because code bases that use pyright for type checking depend on the current inference rules.

**Reply by diepala** [reply]
I see, I understand that it may involve breaking changes. What about adding a third option to the "strictListInference" config, or adding a fully new config option that enables this behaviour? I think this could be beneficial, specially for better autocompletion in VS Code.

**Reply by erictraut** [reply]
No, adding yet more options in this case isn't the right approach. You should use a type declaration if inference doesn't provide the answer you are looking for.

```python
var: list[BaseClass] = [A(), B()]
```

**Reply by diepala** [reply]
Ok, I guess it's fine to make the type explicit in this case.
